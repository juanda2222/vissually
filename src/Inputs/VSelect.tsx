import React, { useContext, useState, useEffect, useRef } from 'react'


import styles from './VSelect.module.css'
import { DefaultThemes, ThemeContext } from "../ThemeProvider/ThemeProvider"
import VButton from "../Buttons/VButton"
import VClickableList from "../Lists/VClickableList"
import ArrowDownSVG from "../assets/arrow-down.js"
import ArrowUpSVG from "../assets/arrow-up.js"


const VSelect: React.FunctionComponent<SelectProps> = ({
  //styling properties
  className,
  style,
  theme,
  primary, secundary, dark,
  listClassName,

  //functional properties
  list,
  onSelect,
  label,
}) => {

  const [isOpen, setOpen] = useState(false)
  const [selectedItem, setSelectedItem] = useState("")


  // get the theme from the context
  const contex_theme: Theme = useContext(ThemeContext);
  const wrapped_className = [styles.button_style, className ? className : ""].join(" ")
  const wrapped_list_className = [
    styles.list_style,
    isOpen ? styles.fadeIn : styles.fadeOut,
    listClassName ? listClassName : ""
  ].join(" ")

  // generate the theme depending on the boolean inputs or the theme input
  var current_theme;
  if (secundary) {
    current_theme = DefaultThemes.secundary
  } else if (dark) {
    current_theme = DefaultThemes.dark
  } else if (primary) {
    current_theme = DefaultThemes.primary
  } else if (!(typeof (contex_theme) == "undefined")) {
    current_theme = contex_theme
  } else {
    current_theme = theme ? theme : DefaultThemes.primary
  }

  // use effect to subscribe the outside click element
  const wrapperRef = useRef<HTMLDivElement>(null);
  useEffect(() => {

    // on outside click handler
    const handleClickOutside = (event: MouseEvent) => {
      if (wrapperRef.current && !wrapperRef.current.contains(event.target as Node)) {
        setOpen(false)
      }
    }
    document.addEventListener('mousedown', handleClickOutside);

    // cleanup:
    return () => {
      // Unbind the event listener on clean up
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [wrapperRef]);

  return (
    <div
      ref={wrapperRef}
      className={styles.big_container}>
      <VButton
        isPressed={isOpen}
        style={style}
        className={wrapped_className}
        theme={current_theme}
        onClick={() => {
          setOpen(!isOpen)
        }}
      >
        
        {selectedItem == "" ? label : selectedItem}
        
        {isOpen ?
          <ArrowUpSVG style={{
            fill: current_theme.textColor2,
            width: "0.7em", height: "0.7em",
            margin: "auto 6px"
          }} /> :
          <ArrowDownSVG style={{
            fill: current_theme.textColor2,
            width: "0.7em", height: "0.7em",
            margin: "auto 6px"
          }} />
        }

      </VButton>
      <VClickableList
        list={list}
        className={styles.list_item}
        containerClassName={wrapped_list_className}
        onClick={(index, value) => {
          setOpen(false)
          setSelectedItem(value)
          onSelect && onSelect(index, value)
        }}
      />
    </div>
  )
}


export default VSelect