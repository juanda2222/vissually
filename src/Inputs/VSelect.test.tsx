
import React from "react"
import renderer from 'react-test-renderer';
import VSelect from './VSelect'

describe('VButton', () => {

  test('Button declaration has no errors', () => {
    expect(VSelect).toBeTruthy()
  })

  test("Default render. Snapshot test", () => { 
    const ButtonTree = renderer.create(<VSelect/>).toJSON();
    console.log(ButtonTree)
    expect(ButtonTree).toMatchSnapshot();
  })

  test("Rendering with props. Snapshot test", () => { 
    
    const ButtonWithProps = renderer.create(
        <VSelect
        list={["some_thing", "some_name", "some_name2", "other name", "whatever list"]}
        onSelect={(i, value) => { console.log("clicked!", i, value) }}
        label = "label"
        />
    ).toJSON();

    console.log(ButtonWithProps)
    expect(ButtonWithProps).toMatchSnapshot();
  })


})
