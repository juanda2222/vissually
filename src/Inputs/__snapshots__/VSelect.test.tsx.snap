// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VSelect Default render. Snapshot test 1`] = `
<div
  className="big_container"
>
  <div
    className="sc-AxjAm gDXVAl action_button button_style "
    onClick={[Function]}
  >
    label
    <svg
      height="451.847px"
      id="Capa_1"
      style={
        Object {
          "fill": "rgb(255, 255, 255)",
          "height": "0.7em",
          "margin": "auto 6px",
          "width": "0.7em",
        }
      }
      version="1.1"
      viewBox="0 0 451.847 451.847"
      width="451.847px"
      x="0px"
      y="0px"
    >
      <path
        d="M225.923,354.706c-8.098,0-16.195-3.092-22.369-9.263L9.27,151.157c-12.359-12.359-12.359-32.397,0-44.751
 c12.354-12.354,32.388-12.354,44.748,0l171.905,171.915l171.906-171.909c12.359-12.354,32.391-12.354,44.744,0
 c12.365,12.354,12.365,32.392,0,44.751L248.292,345.449C242.115,351.621,234.018,354.706,225.923,354.706z"
      />
    </svg>
  </div>
  <div
    className="sc-AxirZ fVduVY list_container list_style fadeOut "
  >
    <div
      className="sc-AxiKw sc-AxhCb gcNYfB list_item list_item"
      onClick={[Function]}
    >
      1
    </div>
    <div
      className="sc-AxiKw sc-AxhUy bjGxvv list_item list_item"
      onClick={[Function]}
    >
      2
    </div>
  </div>
</div>
`;

exports[`VSelect Rendering with props. Snapshot test 1`] = `
<div
  className="big_container"
>
  <div
    className="sc-AxjAm gDXVAl action_button button_style "
    onClick={[Function]}
  >
    label
    <svg
      height="451.847px"
      id="Capa_1"
      style={
        Object {
          "fill": "rgb(255, 255, 255)",
          "height": "0.7em",
          "margin": "auto 6px",
          "width": "0.7em",
        }
      }
      version="1.1"
      viewBox="0 0 451.847 451.847"
      width="451.847px"
      x="0px"
      y="0px"
    >
      <path
        d="M225.923,354.706c-8.098,0-16.195-3.092-22.369-9.263L9.27,151.157c-12.359-12.359-12.359-32.397,0-44.751
 c12.354-12.354,32.388-12.354,44.748,0l171.905,171.915l171.906-171.909c12.359-12.354,32.391-12.354,44.744,0
 c12.365,12.354,12.365,32.392,0,44.751L248.292,345.449C242.115,351.621,234.018,354.706,225.923,354.706z"
      />
    </svg>
  </div>
  <div
    className="sc-AxirZ fVduVY list_container list_style fadeOut "
  >
    <div
      className="sc-AxiKw sc-AxhCb gcNYfB list_item list_item"
      onClick={[Function]}
    >
      some_thing
    </div>
    <div
      className="sc-AxiKw ceVRZF list_item list_item"
      onClick={[Function]}
    >
      some_name
    </div>
    <div
      className="sc-AxiKw ceVRZF list_item list_item"
      onClick={[Function]}
    >
      some_name2
    </div>
    <div
      className="sc-AxiKw ceVRZF list_item list_item"
      onClick={[Function]}
    >
      other name
    </div>
    <div
      className="sc-AxiKw sc-AxhUy bjGxvv list_item list_item"
      onClick={[Function]}
    >
      whatever list
    </div>
  </div>
</div>
`;
